presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -590,6 +606,24 @@ private PlanRoot createLogicalPlanAndOptimize()
         }
     }
 
+    private PlanRoot runCreateLogicalPlanAsync()
+    {
+        try {
+            // Check if creating plan async has been cancelled
+            if (planFutureLocked.compareAndSet(false, true)) {
+                return createLogicalPlanAndOptimize();
+            }
+            return null;
+        }
+        catch (Throwable e) {
+            fail(e);
+            if (e instanceof Error) {
+                throw (Error) e;
+            }
+            throw new GenericInternalException(e);
```

Comment:
why throw a GenericInternalException here?
1. that's really just for presto-common where PrestoExceptions aren't available. Otherwise prefer throwing a PrestoException with appropriate error code.
2. why isn't fail() sufficient to fail the query? And could throwing this exception get in the way here of getting the right error code?

Reply:
Yes, fail() is enough to fail the query. We need to throw again in case `planFuture.get()` is called during`SqlQueryExecution.start()`, then it will stop with the exception. We don't need `GenericInternalException`, I can clean this up.